### ISSUE TITLE
CoffeeScript: "Find All Functions" command finds few functions

### ISSUE THREAD

**Issue by ervumlens** [Komodo/KomodoEdit#376]
(Sorry for the weekend issue dump. At least they're minor.)

*Problem*
Running the "Find All Functions" or "Find Next/Previous Function" in a CoffeeScript file returns a small subset of available functions.

*Example Code* 
The following lines marked "found" are found by the "Find ... Function" commands.

```coffeescript
moduleFunctionNoParams = ->
moduleFunctionWithParams = (i)->
boundModuleFunctionNoParams = =>
boundModuleFunctionWithParams = (i)=>	#found
boundModuleFunctionWithParamsA = boundModuleFunctionWithParamsB = (i) =>
class GenericClass
	constructor: ->
	memberFunctionNoParams: ->
	memberFunctionWithParams: (i) ->
	boundMemberFunctionNoParams: =>
	boundMemberFunctionWithParams: (i) => 	#found
	@classFunctionNoParams: ->
	@classFunctionWithParams: (i) ->
	@boundClassFunctionNoParams: =>
```

The regexp that appears in my finder dialog is as follows:
`^[ |\t]*?(?:([\w|\.|_]*?)\s*[=:]\s*\(.*\)\s*=>).*?$`

This regexp is limited to finding functions that are non-class (no `@` matches), bound functions (no `->` matches) with a parameter spec (no `... = =>` matches).

Two changes to the above regexp produce the following:

`^[ |\t]*?@?(?:([\w|\.|_]*?)\s*[=:]\s*(?:\(.*\))?\s*[=-]>).*?$`

This regexp finds class/non-class, bound/unbound, with/without param spec functions. It does not find functions assigned to multiple variables, e.g. `boundModuleFunctionWithParamsA` and `boundModuleFunctionWithParamsB` in the code example above. 

I don't know where the original regexp is defined, otherwise I'd mention the file name here.

Komodo version: Komodo Edit 9.1.0 build 15798 (platform linux-x86_64, running on Linux 3.13.0-37-generic version #64-Ubuntu SMP Mon Sep 22 21:28:38 UTC 2014)

**Comment by Naatan** [comment]
Oh boy, I was surprised this even exists. It does not exist in IDE, only Edit:

http://docs.activestate.com/komodo/9.1/search.html#search_functions

I'm actually not a fan of this being part of core Komodo, it's very crafty/hacky. Imo this should just be turned into a toolbox tool that we can put in the samples folder. Can incorporate the suggested tweak while we're at it.

Thanks for reporting!

**Comment by ervumlens** [comment]
Yeah, I think it's hacky too, at least as a built-in, so I'm glad we agree. But I figured hey, somebody must want it if it's in there so it better work right. :) 

I thought later that maybe I was running a sample macro the whole time. FWIW, it makes sense to me to put it there.

**Comment by cgchoffman** [comment]
A couple things:
- this is NOT Edit only.  The only thing that is Edit only is the Navigation menu items
- this hooks into the base language IDL file for all supported languages (uses the `namedBlockRE` prop)
- You can't actually remove all of this to a Userscript very easily since it uses the `namedBlockRE` and it appears that the only reason that property exists is to be used in this function.  That regex is the most error prone mechanism in the code in question.

It can of course still be moved to a Userscript so we don't have to maintain the "command" aspect of the functionality but we would have to autodetect the language and manually load the regex whether from a array in the Userscript or from the Language Object.

Thoughts @Naatan?  This is going to be surgery rather than just scooping.
Refs:
https://github.com/Komodo/KomodoEdit/blob/master//src/chrome/komodo/content/library/findController.js#L191
Coffescript lang obj:
https://github.com/Komodo/KomodoEdit/blob/master//src/languages/koJavaScriptLanguage.py#L271

**Comment by cgchoffman** [comment]
Same question: https://github.com/Komodo/KomodoEdit/issues/1741#issuecomment-375124688

**Comment by Naatan** [comment]
Same answer

**Comment by cgchoffman** [comment]
@Naatan I'm looking at Edit.
